### PR TITLE
[Metric] Fix crashed when register metric view in multithread

### PR DESCRIPTION
**Pull request by ashione**
We fond there will be race while registering metric in multithread and process crash finally.
According to opencensus source code, it's known issue but not fixed yet.
```cpp
void StatsManager::ViewInformation::AddConsumer() {
  mu_->AssertHeld();
  ++num_consumers_;
}

int StatsManager::ViewInformation::RemoveConsumer() {
  mu_->AssertHeld();
  return --num_consumers_;
}
```
```cpp
  // TODO: PERF: Global synchronization is only needed for adding or
  // removing measures--we can reduce recording contention by claiming a reader
  // lock on mu_ and a writer lock on a measure-specific mutex.
  mutable absl::Mutex mu_;
```

So we'd better add lock for registering metric when first recording.

## Related issue number
https://github.com/ray-project/ray/issues/13484

## Checks

- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [X] Unit tests
   - [X] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by ashione** [comment]
@zhongchun

**Review by zhongchun — APPROVED** [review]
LGTM.

**Review by simon-mo — APPROVED** [review]

File: src/ray/stats/metric.h
Comment (rkooo567) [inline src/ray/stats/metric.h:133]: Can you rename it to registration_mutex_?

**Review by rkooo567 — COMMENTED** [review]
One minor comment.

**Review by jovany-wang — APPROVED** [review]
Looks good.

File: src/ray/stats/metric.h
Comment (ashione) [reply src/ray/stats/metric.h:133]: > Can you rename it to registration_mutex_?

Nice catch.

**Comment by jovany-wang** [comment]
I guess this can be merged if CI passes.

**Comment by jovany-wang** [comment]
Hi @ashione , should we hold this PR until we fix it totally? If so, I will mark the state to RequestChange.

**Review by jovany-wang — CHANGES_REQUESTED** [review]
@ashione Please ping me if this get fixed totally~

**Review by jovany-wang — APPROVED** [review]
Since there is no failure any more, this can be merged now I think.

**Comment by ashione** [comment]
> Since there is no failure any more, this can be merged now I think.

No related failures to this PR.

**Pull request merged**
